src/query/api/v1/handler/prom/read.go

```diff
@@ -39,15 +39,20 @@ import (
 	"go.uber.org/zap"
 )
 
+// RequestParser is a function that parses request arguments
+type RequestParser func(ctx context.Context, r *http.Request) (models.RequestParams, error)
```

Comment:
If you are planning to have more extension points like this (eg. `ResponseHandler` and so on), might be better to introduce an interface for them, so that you would have to inject a single thing.

Reply:
Good point. Maybe existing `Options` could be used for that?

https://github.com/m3db/m3/blob/dd587475109860fbec289d6912f119caa83cb670/src/query/api/v1/handler/prom/prom.go#L41-L44

E.g. add a field for read request parser and create a function `NewOptions()`, which would set the default parser from the get-go. The latter could be changed on demand.

```go
opts := prom.NewOptions(promEngine)
defaultParser := opts.ReadRequestParser
opts.ReadRequestParser = func(/* ... */) /* ... */ {
  params, err := defaultParser(/* ... */)
  // modify params
  return params, err
}
```

Should be easy to add additional ways to configure the handler in a backwards compatible way. We could also get rid of `NewReadHandlerWithCustomParser` function.

Reply:
Ideally `Options` would be immutable (see other use cases).
I suggest to discuss the approach with @soundvibe .

Reply:
BTW there is an example of what I meant in my suggestion: https://github.com/m3db/m3/blob/2c170ebd9c4acf6f132213eb07bf7c87a1077b14/src/dbnode/storage/types.go#L1393-L1397
Not sure how applicable to your case, but this would be more convenient than function injecting from the point where you have to inject more than a single function.

Reply:
Yes, something like that. But please treat it just as a suggesstion - you know your use case in full picture better than I do.